File: apps/web/src/types/privacy.ts
/**
 * Types for the privacy endpoints (summary, export, account deletion).
 */

export interface PrivacySummary {
  goals: number;
  weight_entries: number;
  body_measurements: number;
  custom_exercises: number;
  workout_templates: number;
  workouts: number;
  custom_foods: number;
  food_logs: number;
  water_logs: number;
  sleep_logs: number;
  daily_steps: number;
  wellness_logs: number;
  habits: number;
}

/** The export is an open, domain-grouped snapshot; we only type its metadata. */
export interface DataExport {
  export_metadata: {
    format_version: string;
    generated_at: string;
    user_id: string;
    email: string;
  };
  [key: string]: unknown;
}

export interface AccountDeletedResponse {
  status: string;
  message: string;
}

/** Human-readable labels for the summary categories, in display order. */
export const SUMMARY_LABELS: { key: keyof PrivacySummary; label: string }[] = [
  { key: "goals", label: "Goals" },
  { key: "workouts", label: "Workouts" },
  { key: "workout_templates", label: "Workout templates" },
  { key: "weight_entries", label: "Weight entries" },
  { key: "body_measurements", label: "Body measurements" },
  { key: "food_logs", label: "Food logs" },
  { key: "custom_foods", label: "Custom foods" },
  { key: "water_logs", label: "Water logs" },
  { key: "sleep_logs", label: "Sleep logs" },
  { key: "daily_steps", label: "Daily steps" },
  { key: "wellness_logs", label: "Wellness check-ins" },
  { key: "habits", label: "Habits" },
  { key: "custom_exercises", label: "Custom exercises" },
];

File: apps/api/app/schemas/privacy.py
"""Pydantic schemas for the privacy endpoints (export, summary, deletion)."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class PrivacySummary(BaseModel):
    """Per-category count of the records a user owns.

    Lets the privacy page show the user exactly what an export or an account
    deletion covers before they commit to either action.
    """

    model_config = ConfigDict(from_attributes=True)

    goals: int = 0
    weight_entries: int = 0
    body_measurements: int = 0
    custom_exercises: int = 0
    workout_templates: int = 0
    workouts: int = 0
    custom_foods: int = 0
    food_logs: int = 0
    water_logs: int = 0
    sleep_logs: int = 0
    daily_steps: int = 0
    wellness_logs: int = 0
    habits: int = 0
    stress_logs: int = 0
    mindfulness_logs: int = 0


class DataExport(BaseModel):
    """Wrapper for a full personal-data export.

    The export groups records by domain. Its shape is intentionally open
    (``dict[str, Any]`` payload) because it mirrors the database and grows with
    the schema; ``export_metadata.format_version`` signals the structure.
    """

    model_config = ConfigDict(extra="allow")

    export_metadata: dict[str, Any]


class AccountDeleteRequest(BaseModel):
    """Body for DELETE /api/v1/privacy/account.

    The password is re-verified server-side so a stolen session cookie alone
    cannot destroy an account.
    """

    password: str = Field(min_length=1, description="Current account password.")


class AccountDeletedResponse(BaseModel):
    """Confirmation returned after a successful account deletion."""

    status: str = "deleted"
    message: str = "Your account and all associated data have been permanently deleted."
